File: app/src/main/java/de/xikolo/controllers/course/ProgressFragment.java
package de.xikolo.controllers.course;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import com.yatatsu.autobundle.AutoBundleField;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import de.xikolo.R;
import de.xikolo.controllers.base.LoadingStatePresenterFragment;
import de.xikolo.models.CourseProgress;
import de.xikolo.models.SectionProgress;
import de.xikolo.presenters.base.PresenterFactory;
import de.xikolo.presenters.course.ProgressPresenter;
import de.xikolo.presenters.course.ProgressPresenterFactory;
import de.xikolo.presenters.course.ProgressView;
import de.xikolo.views.SpaceItemDecoration;

public class ProgressFragment extends LoadingStatePresenterFragment<ProgressPresenter, ProgressView> implements ProgressView {

    public static final String TAG = ProgressFragment.class.getSimpleName();

    @AutoBundleField String courseId;

    private ProgressListAdapter adapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public int getLayoutResource() {
        return R.layout.content_progress;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        RecyclerView recyclerView = view.findViewById(R.id.content_view);

        adapter = new ProgressListAdapter(getActivity());

        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(adapter);

        recyclerView.addItemDecoration(new SpaceItemDecoration(
                0,
                getActivity().getResources().getDimensionPixelSize(R.dimen.card_vertical_margin),
                false,
                new SpaceItemDecoration.RecyclerViewInfo() {
                    @Override
                    public boolean isHeader(int position) {
                        return false;
                    }

                    @Override
                    public int getSpanCount() {
                        return 1;
                    }

                    @Override
                    public int getItemCount() {
                        return adapter.getItemCount();
                    }
                }
        ));
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.refresh, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int videoId = item.getItemId();
        switch (videoId) {
            case R.id.action_refresh:
                presenter.onRefresh();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void setupView(CourseProgress cp, List<SectionProgress> spList) {
        adapter.update(cp, spList);
    }

    @NonNull
    @Override
    protected PresenterFactory<ProgressPresenter> getPresenterFactory() {
        return new ProgressPresenterFactory(courseId);
    }

}

File: app/src/main/java/de/xikolo/controllers/course/ProgressFragment.kt
package de.xikolo.controllers.course

import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.yatatsu.autobundle.AutoBundleField
import de.xikolo.R
import de.xikolo.controllers.base.NetworkStateFragment
import de.xikolo.extensions.observe
import de.xikolo.models.dao.CourseProgressDao
import de.xikolo.viewmodels.course.ProgressViewModel
import de.xikolo.views.SpaceItemDecoration

class ProgressFragment : NetworkStateFragment<ProgressViewModel>() {

    companion object {
        val TAG: String = ProgressFragment::class.java.simpleName
    }

    @AutoBundleField
    lateinit var courseId: String

    private lateinit var adapter: ProgressListAdapter

    override val layoutResource: Int = R.layout.content_progress

    override fun createViewModel(): ProgressViewModel {
        return ProgressViewModel(courseId)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val recyclerView = view.findViewById<RecyclerView>(R.id.content_view)

        adapter = ProgressListAdapter(activity!!)

        val layoutManager = LinearLayoutManager(activity)
        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = adapter

        recyclerView.addItemDecoration(SpaceItemDecoration(
            0,
            activity!!.resources.getDimensionPixelSize(R.dimen.card_vertical_margin),
            false,
            object : SpaceItemDecoration.RecyclerViewInfo {
                override fun isHeader(position: Int): Boolean {
                    return false
                }

                override val spanCount: Int
                    get() = 1

                override val itemCount: Int
                    get() = adapter.itemCount
            }
        ))

        viewModel.sectionProgresses
            .observe(viewLifecycleOwner) { sp ->
                val cp = CourseProgressDao.Unmanaged.find(courseId)
                if (sp.isNotEmpty() && cp != null) {
                    adapter.update(cp, sp)
                    showContent()
                }
            }
    }

}
